src/Corax/Queries/SortingMatches/SortingMatch.Comparers.cs

```diff
@@ -0,0 +1,492 @@
+            {
+                return xItem.Address == null ? 0 : 1;
+            }
+
+            if (xItem.Address == null)
+                return -1;
+            var match = AdvMemory.Compare(xItem.Address + 1, yItem.Address + 1, Math.Min(xItem.Length - 1, yItem.Length - 1));
+            if (match != 0)
+                return match;
+
+            var xItemLengthInBits = (xItem.Length - 1) * 8 - (xItem.Address[0] >> 4);
+            var yItemLengthInBits = (yItem.Length - 1) * 8 - (yItem.Address[0] >> 4);
+            return xItemLengthInBits - yItemLengthInBits;
+        }
+    }
+
+    private struct EntryComparerByTerm : IEntryComparer, IComparer<UnmanagedSpan>
+    {
+        private CompactKeyComparer _cmpTerm;
+        private Lookup<long> _lookup;
+
+        public Slice GetSortFieldName(ref SortingMatch<TInner> match) => match._orderMetadata.Field.FieldName;
+        
+        public void Init(ref SortingMatch<TInner> match)
+        {
+            _lookup = match._searcher.EntriesToTermsReader(match._orderMetadata.Field.FieldName);
+        }
+
+        public void SortBatch(ref SortingMatch<TInner> match, LowLevelTransaction llt, PageLocator pageLocator, Span<long> batchResults, Span<long> batchTermIds,
+            UnmanagedSpan* batchTerms,
+            bool descending)
+        {
+            _lookup.GetFor(batchResults, batchTermIds, long.MinValue);
+            Container.GetAll(llt, batchTermIds, batchTerms, long.MinValue, pageLocator);
+            var indirectComparer = new IndirectComparer<CompactKeyComparer>(batchTerms, new CompactKeyComparer());
+            var indexes = SortByTerms(ref match, batchTermIds, batchTerms, descending, indirectComparer);
+            for (int i = 0; i < indexes.Length; i++)
+            {
+                int bIdx = indexes[i];
+                match._results.Add(batchResults[bIdx]);
+            }
+        }
+
+        private static void MaybeBreakTies<TComparer>(Span<long> buffer, TComparer tieBreaker) where TComparer : struct, IComparer<long>
+        {
+            // We may have ties, have to resolve that before we can continue
+            for (int i = 1; i < buffer.Length; i++)
+            {
+                var x = buffer[i - 1] >> 15;
+                var y = buffer[i] >> 15;
+                if (x != y)
+                    continue;
+
+                // we have a match on the prefix, need to figure out where it ends hopefully this is rare
+                int end = i;
+                for (; end < buffer.Length; end++)
+                {
+                    if (x != (buffer[end] >> 15))
+                        break;
+                }
+
+                buffer[(i - 1)..end].Sort(tieBreaker);
+                i = end;
+            }
+        }
+
+        private static long CopyTermPrefix(UnmanagedSpan item)
+        {
+            long l = 0;
+            Memory.Copy(&l, item.Address + 1 /* skip metadata byte */, Math.Min(6, item.Length - 1));
```

Comment:
Why `6`?

Reply:
We are copying this to a `long`, which has 8 bytes. 
We are using the first 2 bytes as an index 
And 1 bit is reserved.

Then we take the whole thing and threat it as a number and sort it using AVX.

We need to have a separate discussion explaining this in detail.

Reply:
Yeah, things like that would definitely deserve good, clear explanation (written as a design document?). So such a document should be referred to before using, reviewing or changing this mechanisms.